java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -53,6 +53,110 @@ class ExpressionEvaluator(
         currentScope
 
     override fun <T> visit(expr: Expression.ObjectLiteralExpression<T>): Any = expr.value as Any
+
+    override fun <E, T> visit(expr: Expression.FromExpression<E, T>): Any {
+        var sequence = requireNotNull(currentScope.lookup<Any>(expr.source)) {
+            "${expr.source} is null in current scope."
+        }
+
+        if (sequence is List<*>) {
+            sequence = sequence.asSequence()
+        }
+
+        require(sequence is Sequence<*>) {
+            "${expr.source} of type ${sequence::class} cannot be converted to a Sequence"
+        }
+
+        return sequence.map { value ->
+            currentScope.set(expr.iterationVar, value as Any)
```

Comment:
Should this be a mutable scope, or a hierarchy of scopes?

The place where mutable scope could bite us is temporary variables and sub-queries:
```
from p in stuff
let max = Max(from x in p.numbers select x)
select new Thing {
  a: max,
  b: max / 2,
  c: max > 29,
  d: x + 5 // <- that definitely not allowed!
}
```
Clearly the scope where `x` is available is very limited. WDYT?

If you think that in the long term we should do it hierarchically, but you want to do it in a separate PR I'm ok with this as well.

Reply:
It should be a hierarchy (scope chain). Let me do this in a followup since it could be a big change.